### PR TITLE
MFMA bfloat16 support for odd problem size (m- & k-dimension)

### PR DESCRIPTION
**Pull request by rosenrodt**
This PR completes the support for odd sized input matrices.

Running test against fairly recent ROCmSoftwarePlatform/rocBLAS@6f27dfe9 has passed

### CODE REVIEW

File: Tensile/KernelWriterAssembly.py [hunk 1]
```diff
@@ -1539,13 +1539,16 @@ def initKernel(self, kernel, tPA, tPB ):
       self.defineSgpr("PackedSize0", 1)
     if len(kernel["PackedC1IndicesX"]) > 1:
       self.defineSgpr("PackedSize1", 1)
-
-    if kernel["PackSummationDims"]:
-        self.defineSgpr(self.loopCounterName(kernel,self.unrollIdx), 1)
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriterAssembly.py]: This case was removed, is it important?

File: Tensile/KernelWriterAssembly.py [hunk 2]
```diff
@@ -6381,8 +6415,16 @@ def globalReadGuardK(self, kernel, tP):
                     vgpr(zeroVgpr), \
                     "vcc", \
                     "gra += 1 (upper)")
+              # pack 2x registers containing 16-bit each into one 32-bit packed half-dword
+              if (kernel["ProblemType"]["DataType"].isHalf() or \
+                kernel["ProblemType"]["DataType"].isBFloat16()) and \
+                kernel["AssertSummationElementMultiple"] % 2 != 0:
+                if r % 2 == 1 and r > 0:
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriterAssembly.py]: I don't think it's possible for r to be negative here, was this if meant to test something else?

File: Tensile/KernelWriterAssembly.py [hunk 3]
```diff
@@ -7046,10 +7088,18 @@ def localReadDo(self, kernel, bufferIdx, iui, epsi, tP):
...
-
+    loopIdx = self.unrollIdx
+    loopDim = kernel["ProblemType"]["IndicesSummation"][loopIdx]
+    loopChar = self.indexChars[loopDim]
     # mfma: for AB tile in NT layout
     if kernel["MatrixInstruction"] and numReadsAlongK > 1:
       tmpVgpr = self.vgprPool.checkOut(2)
+      maskVgpr = None
+      isLastProductSgpr = self.getTmpSgpr(2)
+      if kernel["MatrixInstruction"] and \
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriterAssembly.py]: There might be a merge conflict here, you should get latest code

File: Tensile/KernelWriterAssembly.py [hunk 4]
```diff
@@ -7409,6 +7473,140 @@ def shiftVectorComponents(self, kernel, tP):
...
+      tmpLabels = []
+      tmp2Labels = []
+      for v in range(0, numSubOutputGroupsPerWave0):
+        label = self.getLabelNum("ShiftVectorComponents%u_R%u_V%u" % (tP["idx"], r, v) )
+        tmpLabels.append(label)
+        tmp2Labels2 = []
+        for o in range(0, numShiftBlock):
+          label = self.getLabelNum("ShiftVectorComponents%u_R%u_V%u_O%u" % (tP["idx"], r, v, o) )
+          tmp2Labels2.append(label)
+        tmp2Labels.append(tmp2Labels2)
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriterAssembly.py]: Is this a typo?

File: Tensile/KernelWriterAssembly.py [hunk 5]
```diff
@@ -7409,6 +7473,140 @@ def shiftVectorComponents(self, kernel, tP):
...
+    numSubOutputGroupsPerWave0 = kernel[tP["mt"]] // numSubOutputPerWave0
+    numShiftBlock              = numContinuousOutput // glvw
+    numOutputElements          = numSubOutputGroupsPerWave0 * numContinuousOutput
+
+    subTile1 = (kernel["MacroTile1"] // ((kernel["SubGroup0"] * kernel["SubGroup1"]) // numThreadInWave)) // kernel["MatrixInstN"]
+
+    # labels
+    svrLabels = []
+    sviLabels = []
+    svoLabels = []
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriterAssembly.py]: Comment defining svr, svi, svo could help

File: Tensile/KernelWriterAssembly.py [hunk 6]
```diff
@@ -7409,6 +7473,140 @@ def shiftVectorComponents(self, kernel, tP):
...
+    divisor = numContinuousOutput
+    kStr += vectorStaticDivide(mReg, wgMT, divisor, tmpVgpr, tmpSgpr)
+    divisor = numOutputThreads0
+    kStr += vectorStaticRemainder(dummy, mReg, mReg, divisor, tmpVgpr, tmpSgpr)
+
+    tReg = self.vgprPool.checkOut(1)
+    divisor = numThreadInWave
+    kStr += vectorStaticRemainder(dummy, tReg, "Serial", divisor, tmpVgpr, tmpSgpr)
+    divisor = kernel["MatrixInstN"]
+    kStr += vectorStaticDivide(tReg, tReg, divisor, tmpVgpr, tmpSgpr)
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriterAssembly.py]: Comments would help out here

File: Tensile/KernelWriterAssembly.py [hunk 7]
```diff
@@ -8464,14 +8662,18 @@ def setupStoreElementsForBatch(self, kernel, gwvw, batchElements, batchElementSg
 
         if self.cfg.numVgprsPerDataPerVI > 0:
           if self.cfg.halfDataRegPerVI:
-            if elementIdx%2 == 0:
-              # allocate for two elements:
+            if kernel["ProblemType"]["DataType"].isBFloat16():
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriterAssembly.py]: Does this change apply to FP16?

File: Tensile/KernelWriterAssembly.py (hunk 4 above)
Comment (TonyYHsieh) [reply Tensile/KernelWriterAssembly.py]: no, we have two level label.

File: Tensile/KernelWriterAssembly.py (hunk 7 above)
Comment (TonyYHsieh) [reply Tensile/KernelWriterAssembly.py]: yes. But we didn't implement FP16 mfma yet.
Anthony will complete FP16 generator later and change it for FP16.

File: Tensile/KernelWriterAssembly.py (hunk 2 above)
Comment (rosenrodt) [reply Tensile/KernelWriterAssembly.py]: Right. `r>0` shouldn't be here. Probably some code left over from earlier testing

File: Tensile/KernelWriterAssembly.py (hunk 1 above)
Comment (bensander) [reply Tensile/KernelWriterAssembly.py]: yes we should keep this code

File: Tensile/KernelWriterAssembly.py (hunk 4 above)
Comment (TonyYHsieh) [reply Tensile/KernelWriterAssembly.py]: update comment and make a slide to explain in e-mail

File: Tensile/KernelWriterAssembly.py (hunk 5 above)
Comment (TonyYHsieh) [reply Tensile/KernelWriterAssembly.py]: update comment and make a slide to explain in e-mail

File: Tensile/KernelWriterAssembly.py (hunk 6 above)
Comment (TonyYHsieh) [reply Tensile/KernelWriterAssembly.py]: update comment and make a slide to explain in e-mail

File: Tensile/KernelWriterAssembly.py (hunk 3 above)
Comment (rosenrodt) [reply Tensile/KernelWriterAssembly.py]: latest code pushed

File: Tensile/KernelWriterAssembly.py (hunk 1 above)
Comment (rosenrodt) [reply Tensile/KernelWriterAssembly.py]: my mistake. they are now back in latest push

**Comment by rosenrodt** [comment]
Some errors running on gfx900 in the CI:
```
Index:  Device | Reference
[0]  elem=0 idx=0: -nan!=-54
[1]  elem=1 idx=1: -nan!=12
[2]  elem=2 idx=2: -nan!=82
[3]  elem=3 idx=3: -nan!=56

Performed bounds check on 512 elements (512 before data)
BOUNDS CHECK:
Before buffer: found 0 errors in 512 values checked.
Inside buffer: found 0 errors in 0 values checked.
After buffer: found 0 errors in 512 values checked.

0, 24/25, 0/35, Contraction_l_Ailk_Bljk_Cijk_Dijk, "(130,130,2,66)", Cijk_Ailk_Bljk_HBH_MT64x32x16_SE_AF0EM1_ASEM1_GRVW2_K1_LDL1_TT4_2_VW2_WG16_16_1, FAILED, 31.56, 141, 25.00, 852.00, 960.00, 945.00, 0, 1, 2020-02-18 11:34:02.695625
```

The kernel candidate, if run on gfx906/gfx908, will be rejected because it satisfy the condition `HasEccHalf`, which will prevent the kernel from even being tested against. So this bug slipped through since we have yet to test on gfx900. Could you clarify a bit what implications `HasEccHalf` have?

File: Tensile/KernelWriterAssembly.py [hunk 8]
```diff
@@ -7116,10 +7149,25 @@ def localReadDo(self, kernel, bufferIdx, iui, epsi, tP):
 
             valuIdx += blockWidth
 
-            if kIdx % 2 == 1 and kIdx > 0:
+            if kIdx % 2 == 1:
               # pack 2 half-dword into one
               destVgpr = vgpr("Valu%s_X%u_I%u+%u" % (tc, bufferIdx, iui, rIdx))
               imod.addInst("s_waitcnt lgkmcnt(0)", "")
+              if kernel["MatrixInstruction"] and \
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriterAssembly.py]: "MatrixInstruction" was already checked above

Comment (rosenrodt) [reply Tensile/KernelWriterAssembly.py]: Will clean that up in the next PR. Thanks!

**Pull request merged**